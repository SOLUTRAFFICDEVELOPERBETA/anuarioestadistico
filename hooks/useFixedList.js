/* eslint-disable indent */

import React from 'react';
import { List } from '@material-ui/core';
import PropTypes from 'prop-types';
import Spinner from '../components/Spinner';
import PaginationBar from '../components/PaginationBar';

/**
 * Hook para mostrar una lista limitada.
 * @param {*} step Tamaño del arreglo
 */
const useFixedList = (step = 5) => {
    const [total, setTotal] = React.useState([]);
    const [actual, setActual] = React.useState([]);
    const [position, setPosition] = React.useState(0);
    const [maxPages, setMaxPages] = React.useState(1);
    const [load, setLoad] = React.useState(true);

    const onSkipNext = () => {
        setPosition(maxPages - 1);
        setLoad(true);
    };

    // Método para avanzar al siguiente grupo de la lista
    const onNext = () => {
        const size = position * step + step;

        if (size > total.length) {
            setPosition(0);
            return;
        }

        setPosition(position + 1);
        setLoad(true);
    };

    const onSkipBack = () => {
        setPosition(0);
        setLoad(true);
    };

    // Método para retroceder en la lista
    const onBack = () => {
        const size = position * step - step;

        if (size < 0) {
            setPosition(0);
            return;
        }

        setPosition(position - 1);
        setLoad(true);
    };

    /**
     * Componente a mostrar de la lista limitado
     * @param {{ itemRender }} props Propiedades del componente
     */
    const Container = ({ itemRender }) => (
        <>
            {load ? <Spinner /> : <List>{actual.map((item) => itemRender(item))}</List>}
            {maxPages > 0 ? (
                <PaginationBar
                    onBack={onBack}
                    onSkipBack={onSkipBack}
                    showBack={position === 0}
                    onNext={onNext}
                    onSkipNext={onSkipNext}
                    showNext={position === maxPages - 1}
                    label={`Pagina ${position + 1} de ${maxPages}`}
                />
            ) : null}
        </>
    );

    Container.propTypes = {
        itemRender: PropTypes.any.isRequired
    };

    React.useEffect(() => {
        if (position > maxPages) {
            setPosition(0);
        }
    }, [position, maxPages]);

    React.useEffect(() => {
        const max = Math.ceil(total.length / step);
        setMaxPages(max);
    }, [total, step]);

    React.useEffect(() => {
        const frag = [...total];
        setActual(frag.splice(position * step, step));
        setTimeout(() => {
            setLoad(false);
        }, 500);
    }, [total, position, step]);

    return [setTotal, Container, [...total]];
};

export default useFixedList;
