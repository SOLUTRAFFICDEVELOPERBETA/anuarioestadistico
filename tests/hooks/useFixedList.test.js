import '@testing-library/jest-dom';
import { renderHook, act } from '@testing-library/react-hooks';
import { mount } from 'enzyme';
import useFixedList from '../../hooks/useFixedList';
import { itemList } from '../fixtures';

describe('Pruebas en useFixedList', () => {

    test('debe retornar un componente de tipo container, sin la lista ', async () => {
        const { result, waitForNextUpdate } = renderHook(() => useFixedList(10));
        const itemRender = jest.fn();

        const [setTotal, Container] = result.current;
        await waitForNextUpdate()
        expect(typeof setTotal).toBe('function');
        const wrapper = mount(<Container itemRender={itemRender} />);
        expect(wrapper.find('div').at(1).hasClass('spinner')).toBe(true)

    });

    test('debe cargar la lista de elementos envÃ­ada', async () => {
        const { result, waitForNextUpdate } = renderHook(() => useFixedList(10));

        const itemRender = jest.fn();
        const [setTotal] = result.current;
        await waitForNextUpdate()
        act(() => {
            setTotal(itemList);
        });
        const [, Container, total] = result.current;

        const wrapper = mount(<Container itemRender={itemRender} />);
        expect(wrapper).toMatchSnapshot();
        expect(total.length).toBe(10);
        expect(wrapper.hasClass('spinner')).toBe(false);
        expect(wrapper.find('span').at(0).text().trim()).toBe('Pagina 1 de 1')
        expect(wrapper.find('button').exists()).toBe(true)
    })


})