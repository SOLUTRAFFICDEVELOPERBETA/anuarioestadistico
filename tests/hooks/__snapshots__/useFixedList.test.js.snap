// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en useFixedList debe cargar la lista de elementos env√≠ada 1`] = `
Array [
  <ul
    className="MuiList-root MuiList-padding"
  />,
  <div
    className="MuiBox-root MuiBox-root-1"
  >
    <div
      className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-align-content-xs-center MuiGrid-justify-xs-flex-end"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto MuiGrid-grid-sm-6 MuiGrid-grid-md-3 MuiGrid-grid-lg-3 MuiGrid-grid-xl-3"
      >
        <span
          className="MuiTypography-root MuiTypography-caption MuiTypography-colorPrimary MuiTypography-alignCenter MuiTypography-displayBlock"
        >
          Pagina 1 de 1
        </span>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-auto MuiGrid-grid-sm-6 MuiGrid-grid-md-2 MuiGrid-grid-lg-2 MuiGrid-grid-xl-2"
      >
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 6h2v12H6zm3.5 6l8.5 6V6z"
              />
            </svg>
          </span>
        </button>
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </span>
        </button>
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </span>
        </button>
        <button
          className="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary Mui-disabled MuiIconButton-sizeSmall Mui-disabled"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <span
            className="MuiIconButton-label"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M6 18l8.5-6L6 6v12zM16 6v12h2V6h-2z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>,
]
`;
